### ISSUE TITLE
[8.1] Incorrect filesize uploading files via Webdav as recipient using encryption

### ISSUE THREAD

**Issue by xn--nding-jua** [owncloud/core#17594]
### Steps to reproduce
1. Upload a file using webdav (e.g. FileSync on Android) - here a 1.4 MB file - using the mode "use temporary files"
2. The uploaded file will be encrypted, put into the ownCloud filesystem and the filesize within the oc_filecache-database is set to the size of the encrypted file (1.9 MB). The flag "encrypted" will be not set within oc_filecache
3. Using the web-interface the JPG-file can be displayed (1.4MB) and downloaded (1.4MB) but the ownCloud-Windows-Client aborts with errors and the file cannot be downloaded using a browser and the direct webdav-link (the browser tries to download 1.9MB and aborts after 1.4MB)

### Expected behaviour
The filesize should be published to the webdav-clients correctly. The decrypted filesize has to be stored within oc_filecache and the flag "encrypted" has to be set to 1.

### Actual behaviour
The stored filesize is the size of the encrypted file so that ownCloud-Windows-Client and webdav-clients cannot operate any more. Setting up the correct (decrypted) filesize within the database fixes the problem immediatly!

The problem is not there on some files if you are using the Webinterface, because clicking on a file will decrypt the file to the temporary folder immediatly so that the correct file size will be published to the browser.

The problem seems to occure, if the webdav-client creates a temporary file for upload-process and moves this file after the upload completes (like FileSync does in default settings). Maybe this upload-encrypt-move-process has a bug!?

### Server configuration
**Operating system**:
Ubuntu 14.04.2 LTS

**Web server:**
Apache (unknown version)

**Database:**
SQL 5.4

**PHP version:**
PHP 5.5

**ownCloud version:** (see ownCloud admin page)
8.1.0.8

**Updated from an older ownCloud or fresh install:**
from an older installation (8.0.2)

**List of activated apps:**
 activity: 2.0.1
 - bookmarks: 0.4
 - calendar: 0.7.2
 - contacts: 0.4.0.0
 - documents: 0.10.0
 - encryption: 1.0.0
 - files: 1.1.9
 - files_gpxviewer_extended: 1.4
 - files_locking:
 - files_pdfviewer: 0.7
 - files_sharing: 0.6.2
 - files_texteditor: 0.4
 - files_trashbin: 0.6.3
 - files_versions: 1.0.6
 - files_videoviewer: 0.1.3
 - firstrunwizard: 1.1
 - galleryplus: 13
 - mail: 0.1.3
 - music: 0.3.6
 - tasks: 0.6.0
 - updater: 0.6
Disabled:
 - external
 - files_external
 - gallery
 - ocsms
 - provisioning_api
 - templateeditor
 - user_external
 - user_ldap
 - user_webdavauth

**The content of config/config.php:**

<?php
$CONFIG = array (
  'instanceid' => 'xxxx',
  'passwordsalt' => 'xxxx',
  'trusted_domains' => 
  array (
    0 => 'www.xxxx.de',
    1 => 'cloud.xxxx.de',
    2 => 'xxxx.de',
  ),
  'datadirectory' => '/www/htdocs/xxxx/xxxx/cloud/data',
  'dbtype' => 'mysql',
  'version' => '8.1.0.8',
  'dbname' => 'xxxx',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'xxxx',
  'dbpassword' => 'xxxx',
  'installed' => true,
  'theme' => '',
  'maintenance' => false,
  'mail_from_address' => 'xxxx',
  'mail_smtpmode' => 'php',
  'mail_domain' => 'xxxx.de',
  'forcessl' => true,
  'trashbin_retention_obligation' => 60,
  'trashbin_auto_expire' => true,
  'loglevel' => 2,
  'log_rotate_size' => false,
  'logtimezone' => 'Europe/Berlin',
  'check_for_working_webdav' => true,
  'check_for_working_htaccess' => true,
  'xframe_restriction' => false,
  'secret' => 'xxxx',
  'forceSSLforSubdomains' => true,
  'tempdirectory' => '/www/htdocs/xxxx/tmp',
  'enable_previews' => true,
  'enabledPreviewProviders' => 
  array (
    0 => 'OC\\Preview\\Image',
    1 => 'OC\\Preview\\MP3',
    2 => 'OC\\Preview\\TXT',
    3 => 'OC\\Preview\\MarkDown',
    4 => 'OC\\Preview\\Illustrator',
    5 => 'OC\\Preview\\Postscript',
    6 => 'OC\\Preview\\Photoshop',
    7 => 'OC\\Preview\\TIFF',
  ),
  'singleuser' => false,
);

**Are you using external storage, if yes which one:** local/smb/sftp/...
no

**Are you using encryption:** yes/no
yes

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
no

### Client configuration
ownCloud Client 1.8.3
**Browser:**
Firefox 39

**Operating system:**
Windows 7 64-Bit

**Comment by karlitschek** [comment]
@schiesbn Can you have a look please?

**Comment by PVince81** [comment]
@xn--nding-jua thanks for the detailed infos.

Do you have an idea about how the "use temporary file" mode works in that Filesync app you use ? What naming scheme is used for the temporary files ? Do they have the ".part" extension ?

Does Webdav upload with another client work ? Can the ownCloud sync client properly upload files ?
This is to find out whether FileSync is doing something special.

**Comment by xn--nding-jua** [comment]
Hi,

yes "FolderSync" uses temporary files with the ending *.tacitpart. It uploads the file and afterwards moves it to the desired name. By now I've deactivated this function and will check if this avoids the filesize- and encrypted=0 problem within oc_filebase.

Uploaded files with Total Commander seem to work - I will try to rename some files using the webdav-interface. Maybe only the renaming is the problem - or maybe the upload-rename-timing.

The ownCloud-Sync-Client does not seem to be affected except that the Client is useless once the filesize is corrupted. The Client won't work until the file with corrupted filesize is deleted or manually fixed within the sql-database.

When I have more information I will post them here. Around 80 files have been affected yesterday. After I fixed them manually everything went back to normal.

cu,
Chris

**Comment by xn--nding-jua** [comment]
Hi,

again the filesize of more than 70 files are corrupted after a synchronisation with FolderSync Lite on the Android-Device. For now I will deactivate the sync-functions on the Android-devices until this bug will be fixed. During this synchronizations the "temporary-files"-option had been disabled - so FolderSync is using the standard WebDAV connection of ownCloud.

It would be a disaster if this is a general bug that could occure with the ownCloud client! Please have a look at this or maybe give me a hint where I could start to look for the problem?

kind regards,
Chris

**Comment by schiesbn** [comment]
I just tried to reproduce it with 8.1.0. After uploading a file with a normal webdav client the file cache contains the correct un-encrypted size. I also tried it to upload a file with FolderSync, the file also contain the correct unencrypted size in my file cache. I don't know if "use temporary file" is enabled, I can't find the settings but i didn't change the settings so it is probably enabled.

**Comment by xn--nding-jua** [comment]
Hi,

in the meantime the problem occured with the ownCloud Client, too. So this is not a pure FolderSync problem anymore. I've uploaded a MatLAB-figure file - so it is not related to JPG-files alone.

The uploaded and encrypted file has a size of 104 KB on my storage and the ownCloud-Webinterface says it has a size of 104 kB, too. This is wrong as the size of the file on my local computer is only 72 kB.

It has to do something with the timing between the upload-process, the encryption and the calculation of the filesize.

kind regards,
Chris

**Comment by schiesbn** [comment]
Hm, we remember the unencrypted size on fclose. That's the earliest point in time where we know the complete unencrypted size: https://github.com/owncloud/core/blob/master/lib/private/files/stream/encryption.php#L401

Afterwards the file scanner which creates the file cache entry should re-use this unencrypted size: https://github.com/owncloud/core/blob/master/lib/private/files/storage/wrapper/encryption.php#L128

@icewind1991 Do you see a potential timing issue here? Can it happen that the file cache creates the entry before the array with the unencrypted size was updated?

**Comment by xn--nding-jua** [comment]
Hi,

the column "unencrypted_size" is empty for most files in my database:
![oc_filecache](https://cloud.githubusercontent.com/assets/9845353/8801365/57a4962e-2fba-11e5-8849-90fa95eb2b0a.png)

And there, where a filesize is given for a file the column "size" and "encrypted_size" have the same value. Maybe this is the problem? But why are all files working then?

**Comment by schiesbn** [comment]
@xn--nding-jua the "unencrypted_size" column is no longer used in 8.1. We store the unencrypted size directly in the "size" column

**Comment by xn--nding-jua** [comment]
OK. What about the Line 348? You are calculating the max() of $this.unencryptedSize and $this.position. The encrypted files are written in the Flush() function in https://github.com/owncloud/core/blob/master/lib/private/files/stream/encryption.php#L417 and the max() in the stream_wirte() function maybe somehow store the encrypted filesize into the unencryptedSize-storage?

**Comment by PVince81** [comment]
I got a similar issue today.
My wife uploaded a png file with the ownCloud Android client from a tablet.
At the same time, I believe that I was browsing the same folder on my mobile phone using the Android client too. Then the file she just uploaded has the encrypted size instead of unencrypted size in the database.

I suspect this might have happened because a concurrent scanning process has started before the upload finished. The scanner would find the new encrypted file before the file cache / index was updated, so it will believe that it's a new file, and would just write its real size in the DB.
This should be avoidable in the future with the transactional locking.

Still, I think the encryption app should auto-repair the file size in the database. If at the time a file is being streamed to the client it turns out that the actual size did not match the content-length, the encryption app should automatically adjust the cache entry with the correct size.

The current behavior is that it still keeps the wrong size. This causes many clients (Android client and cadaver) to abort because there is no more data coming in.

I'll see if I can reproduce the issue consistently.

**Comment by PVince81** [comment]
Here is the entry in question:
```
+--------+---------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+--------+-------------------------------+----------+----------+--------+------------+-----------+----------------------------------+------------------+---------------+-------------+
| fileid | storage | path                                                                                                                                                             | path_hash                        | parent | name                          | mimetype | mimepart | size   | mtime      | encrypted | etag                             | unencrypted_size | storage_mtime | permissions |
+--------+---------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+--------+-------------------------------+----------+----------+--------+------------+-----------+----------------------------------+------------------+---------------+-------------+
|  62475 |       1 | files/clientsync/petry/pics/colorfy_tingting/sharingCFGFF1437910038590.png                                                                                       | 81abaed2083a73508d1d74293587dab4 |  62309 | sharingCFGFF1437910038590.png |       16 |        6 | 757684 | 1
```
The real file size is 560463, not 757684.
So downloading the file with cadaver will stop at 70%, then it waits for a while then gives up: download failed.

**Comment by PVince81** [comment]
Looks like the `encrypted` flag is also set to 0 even though the file is encrypted, looks like a similar issue like @xn--nding-jua.

This is ownCloud 8.1.0.

@schiesbn what happens when the file scanner finds a non-indexes encrypted file on-disk ? I guess it doesn't try to automatically detect whether it's encrypted ?

**Comment by PVince81** [comment]
A possible workaround for this: set `filesystem_check_changes` to `0` in `config.php`. This assumes that the FS is only used exclusively through ownCloud.

My setting is currently set to 1 (the default), I'll leave it so in the hope to find reproduction steps.

**Comment by PVince81** [comment]
Additional detail, not sure if relevant: the folder "petry" is shared by me with my wife. So she uploaded the file as recipient.

@xn--nding-jua in your case I guess these were not shared files/folders, right ?

**Comment by xn--nding-jua** [comment]
Hi,

no, the folder where the problem occurs is shared with my wife, too. Maybe there is a association here: Encryption-App <-> SharedWithWife? (just kidding *g*)

I will try to deactivate this "filesystem_check_changes" as 20 files are corrupted since last night, again.

kind regards,
Chris

**Comment by PVince81** [comment]
@xn--nding-jua let me know if setting that flag to 0 fixed the issue for you. This would be a clue that indeed it is the scanner checking for new files.

**Comment by xn--nding-jua** [comment]
Hi @PVince81, the last two days the problem did not came up again. I've deactivated the <code>filesystem_check_changes</code> in the config-file as you've suggested. If the problem will come up again, I will post it here. Thank you for the hint!

**Comment by PVince81** [comment]
@xn--nding-jua thanks for the info. This would confirm the "concurrent scanner" theory then. :smile:

**Comment by PVince81** [comment]
This ticket is about making this config set to 0 by default: https://github.com/owncloud/core/issues/16897

**Comment by PVince81** [comment]
I happened again on my env with newly uploaded pics. Note that I didn't disable "filesystem_check_changes" yet:
```
MariaDB [owncloud]> select * from oc_filecache where path like 'files/%' and encrypted=0 and mimetype != 1;
+--------+---------+----------------------------------------------------------------------------+----------------------------------+--------+-------------------------------+----------+----------+--------+------------+-----------+----------------------------------+------------------+---------------+-------------+
| fileid | storage | path                                                                       | path_hash                        | parent | name                          | mimetype | mimepart | size   | mtime      | encrypted | etag                             | unencrypted_size | storage_mtime | permissions |
+--------+---------+----------------------------------------------------------------------------+----------------------------------+--------+-------------------------------+----------+----------+--------+------------+-----------+----------------------------------+------------------+---------------+-------------+
|  62462 |       1 | files/clientsync/petry/pics/colorfy_tingting/sharingCFGFF1437863614199.png | d86fa5ca734969d1d058030d0bba4d4d |  62309 | sharingCFGFF1437863614199.png |       16 |        6 | 555064 | 1437863658 |         0 | 7e6a8a2ced170b0b0403902e8a3136ab |                0 |    1437863658 |          27 |
|  62475 |       1 | files/clientsync/petry/pics/colorfy_tingting/sharingCFGFF1437910038590.png | 81abaed2083a73508d1d74293587dab4 |  62309 | sharingCFGFF1437910038590.png |       16 |        6 | 757684 | 1437910063 |         0 | d2a3a0b7d52a95f20efb0b9ae789bf89 |                0 |    1437910063 |          27 |
|  63085 |       1 | files/clientsync/petry/pics/colorfy_tingting/sharingCFGFF1438189473839.png | 61a8b91dca0bd415d9d2316d2bc1981f |  62309 | sharingCFGFF1438189473839.png |       16 |        6 | 555064 | 1438189500 |         0 | aafc93c70ea7a00f799b21c33ed23306 |                0 |    1438189500 |          27 |
|  63098 |       1 | files/clientsync/petry/pics/colorfy_tingting/sharingCFGFF1438189512673.png | 2101961fa54d55a9097a2ad3e2175bd8 |  62309 | sharingCFGFF1438189512673.png |       16 |        6 | 757684 | 1438189540 |         0 | 502442f5f6fee423f14f03dc44583676 |                0 |    1438189540 |          27 |
|  63111 |       1 | files/clientsync/petry/pics/colorfy_tingting/sharingCFGFF1438189563343.png | 7428427a1f9f9eb2f4ef8a3f0e1b2a0d |  62309 | sharingCFGFF1438189563343.png |       16 |        6 | 612368 | 1438189600 |         0 | 02b885dd1c582604e26f8cdd5ddcf0c7 |                0 |    1438189600 |          27 |
+--------+---------+----------------------------------------------------------------------------+----------------------------------+--------+-------------------------------+----------+----------+--------+------------+-----------+----------------------------------+------------------+---------------+-------------+
5 rows in set (0.03 sec)
```

Now I'll disable filesystem_check_changes. This will help confirm that we are talking about the same bug.

**Comment by PVince81** [comment]
Here is the access log from the time these files were uploaded:
```
[29/Jul/2015:19:04:40 +0200] "HEAD /remote.php/webdav/ HTTP/1.1" 200 - "-" "Android-ownCloud"
[29/Jul/2015:19:04:40 +0200] "HEAD /remote.php/webdav/sharingCFGFF1438189473839.png HTTP/1.1" 404 - "-" "Android-ownCloud"
[29/Jul/2015:19:04:40 +0200] "PUT /remote.php/webdav/sharingCFGFF1438189473839.png HTTP/1.1" 201 - "-" "Android-ownCloud"
[29/Jul/2015:19:04:44 +0200] "PROPFIND /remote.php/webdav/sharingCFGFF1438189473839.png HTTP/1.1" 207 556 "-" "Android-ownCloud"
[29/Jul/2015:19:04:56 +0200] "HEAD /remote.php/webdav/petry/pics/colorfy_tingting/ HTTP/1.1" 200 - "-" "Android-ownCloud"
[29/Jul/2015:19:04:56 +0200] "HEAD /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189473839.png HTTP/1.1" 404 - "-" "Android-ownCloud"
[29/Jul/2015:19:04:57 +0200] "PUT /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189473839.png HTTP/1.1" 201 - "-" "Android-ownCloud"
[29/Jul/2015:19:05:01 +0200] "PROPFIND /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189473839.png HTTP/1.1" 207 584 "-" "Android-ownCloud"
[29/Jul/2015:19:05:18 +0200] "HEAD /remote.php/webdav/ HTTP/1.1" 200 - "-" "Android-ownCloud"
[29/Jul/2015:19:05:18 +0200] "HEAD /remote.php/webdav/sharingCFGFF1438189512673.png HTTP/1.1" 404 - "-" "Android-ownCloud"
[29/Jul/2015:19:05:18 +0200] "PUT /remote.php/webdav/sharingCFGFF1438189512673.png HTTP/1.1" 201 - "-" "Android-ownCloud"
[29/Jul/2015:19:05:23 +0200] "PROPFIND /remote.php/webdav/sharingCFGFF1438189512673.png HTTP/1.1" 207 556 "-" "Android-ownCloud"
[29/Jul/2015:19:05:35 +0200] "HEAD /remote.php/webdav/petry/pics/colorfy_tingting/ HTTP/1.1" 200 - "-" "Android-ownCloud"
[29/Jul/2015:19:05:35 +0200] "HEAD /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189512673.png HTTP/1.1" 404 - "-" "Android-ownCloud"
[29/Jul/2015:19:05:36 +0200] "PUT /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189512673.png HTTP/1.1" 201 - "-" "Android-ownCloud"
[29/Jul/2015:19:05:41 +0200] "PROPFIND /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189512673.png HTTP/1.1" 207 584 "-" "Android-ownCloud"
[29/Jul/2015:19:06:17 +0200] "HEAD /remote.php/webdav/ HTTP/1.1" 200 - "-" "Android-ownCloud"
[29/Jul/2015:19:06:17 +0200] "HEAD /remote.php/webdav/sharingCFGFF1438189563343.png HTTP/1.1" 404 - "-" "Android-ownCloud"
[29/Jul/2015:19:06:17 +0200] "PUT /remote.php/webdav/sharingCFGFF1438189563343.png HTTP/1.1" 201 - "-" "Android-ownCloud"
[29/Jul/2015:19:06:22 +0200] "PROPFIND /remote.php/webdav/sharingCFGFF1438189563343.png HTTP/1.1" 207 556 "-" "Android-ownCloud"
[29/Jul/2015:19:06:36 +0200] "HEAD /remote.php/webdav/petry/pics/colorfy_tingting/ HTTP/1.1" 200 - "-" "Android-ownCloud"
[29/Jul/2015:19:06:36 +0200] "HEAD /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189563343.png HTTP/1.1" 404 - "-" "Android-ownCloud"
[29/Jul/2015:19:06:36 +0200] "PUT /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189563343.png HTTP/1.1" 201 - "-" "Android-ownCloud"
[29/Jul/2015:19:06:40 +0200] "PROPFIND /remote.php/webdav/petry/pics/colorfy_tingting/sharingCFGFF1438189563343.png HTTP/1.1" 207 584 "-" "Android-ownCloud"
[29/Jul/2015:19:07:16 +0200] "GET /status.php HTTP/1.1" 200 95 "-" "Android-ownCloud"
[29/Jul/2015:19:07:16 +0200] "PROPFIND /remote.php/webdav/ HTTP/1.1" 207 558 "-" "Android-ownCloud"
[29/Jul/2015:19:07:16 +0200] "PROPFIND /remote.php/webdav/ HTTP/1.1" 207 8232 "-" "Android-ownCloud"
[29/Jul/2015:19:07:17 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2F&reshares=false&subfiles=true HTTP/1.1" 200 3406 "-" "Android-ownCloud"
[29/Jul/2015:19:07:23 +0200] "GET /status.php HTTP/1.1" 200 95 "-" "Android-ownCloud"
[29/Jul/2015:19:07:23 +0200] "PROPFIND /remote.php/webdav/ HTTP/1.1" 207 558 "-" "Android-ownCloud"
[29/Jul/2015:19:07:23 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2F&reshares=false&subfiles=true HTTP/1.1" 200 3406 "-" "Android-ownCloud"
[29/Jul/2015:19:07:25 +0200] "PROPFIND /remote.php/webdav/petry/ HTTP/1.1" 207 564 "-" "Android-ownCloud"
[29/Jul/2015:19:07:25 +0200] "PROPFIND /remote.php/webdav/petry/ HTTP/1.1" 207 1392 "-" "Android-ownCloud"
[29/Jul/2015:19:07:25 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2Fpetry%2F&reshares=false&subfiles=true HTTP/1.1" 200 127 "-" "Android-ownCloud"
[29/Jul/2015:19:07:26 +0200] "PROPFIND /remote.php/webdav/petry/pics/ HTTP/1.1" 207 569 "-" "Android-ownCloud"
[29/Jul/2015:19:07:26 +0200] "PROPFIND /remote.php/webdav/petry/pics/ HTTP/1.1" 207 6602 "-" "Android-ownCloud"
[29/Jul/2015:19:07:26 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2Fpetry%2Fpics%2F&reshares=false&subfiles=true HTTP/1.1" 200 1464 "-" "Android-ownCloud"
[29/Jul/2015:19:07:27 +0200] "PROPFIND /remote.php/webdav/petry/pics/colorfy_tingting/ HTTP/1.1" 207 583 "-" "Android-ownCloud"
[29/Jul/2015:19:07:27 +0200] "PROPFIND /remote.php/webdav/petry/pics/colorfy_tingting/ HTTP/1.1" 207 4462 "-" "Android-ownCloud"
[29/Jul/2015:19:07:27 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2Fpetry%2Fpics%2Fcolorfy_tingting%2F&reshares=false&subfiles=true HTTP/1.1" 200 127 "-" "Android-ownCloud"
[29/Jul/2015:19:07:35 +0200] "PROPFIND /remote.php/webdav/petry/pics/colorfy_tingting/ HTTP/1.1" 207 583 "-" "Android-ownCloud"
[29/Jul/2015:19:07:35 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2Fpetry%2Fpics%2Fcolorfy_tingting%2F&reshares=false&subfiles=true HTTP/1.1" 200 127 "-" "Android-ownCloud"
[29/Jul/2015:19:07:39 +0200] "PROPFIND /remote.php/webdav/petry/pics/ HTTP/1.1" 207 569 "-" "Android-ownCloud"
[29/Jul/2015:19:07:39 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2Fpetry%2Fpics%2F&reshares=false&subfiles=true HTTP/1.1" 200 1464 "-" "Android-ownCloud"
[29/Jul/2015:19:07:40 +0200] "PROPFIND /remote.php/webdav/petry/ HTTP/1.1" 207 564 "-" "Android-ownCloud"
[29/Jul/2015:19:07:41 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2Fpetry%2F&reshares=false&subfiles=true HTTP/1.1" 200 127 "-" "Android-ownCloud"
[29/Jul/2015:19:07:42 +0200] "GET /status.php HTTP/1.1" 200 95 "-" "Android-ownCloud"
[29/Jul/2015:19:07:42 +0200] "PROPFIND /remote.php/webdav/ HTTP/1.1" 207 558 "-" "Android-ownCloud"
[29/Jul/2015:19:07:42 +0200] "PROPFIND /remote.php/webdav/ HTTP/1.1" 207 8232 "-" "Android-ownCloud"
[29/Jul/2015:19:07:42 +0200] "GET /ocs/v1.php/apps/files_sharing/api/v1/shares?path=%2F&reshares=false&subfiles=true HTTP/1.1" 200 3406 "-" "Android-ownCloud"
```

Strangely I don't see any concurrency.
The sequence seems to be distinct and happen in sequence, not in parallel:
1) Android client does HEAD on remote file to make sure it doesn't exist, server returns 404
2) Client does a PUT of the file
3) Client does a PROPFIND on the file, probably to get more details about it. From here it looks like the PUT operation was already finished (according to http://webmasters.stackexchange.com/questions/65566/when-does-apache-log-to-access-log-before-or-after-serving-the-request Apache only logs a request once it is already finished, but with the timestamp from its start)
4) Repeat steps with the next file

While the "PROPFIND" would trigger the file scanner, at that time the file should already be indexed. So this shouldn't be a concurrency case.

Needs further research and testing.

**Comment by PVince81** [comment]
Ok, it's easy to reproduce:

### Steps to reproduce

1. Enable encryption
1. Create two users "user1" and "user2"
1. Login as "user1"
1. Create a folder "test"
1. Share "test" with "user2"
1. Login as "user2" with a Webdav client (I used cadaver)
1. PUT a bigger file "pic.png" into "test"
1. Check the database entry for "pic.png"
1. Download "pic.png"

### Expected result
Encrypted flag must be set after upload.
File can be downloaded without problem.

### Actual result
Encrypted flag is missing, file size is the one from the encrypted file (happens on subsequent PROPFIND due to the missing encrypted flag)
Downloading the file can hang and time out because some clients wait for extra data because of the wrong content-length.

### Versions
Observed in v8.1.0 and also on the latest stable8.1 and upcoming 8.1.1 (a9524e250a1f53a20a68acd2449e90fa2d13ef51)

Marking as regression + critical.

@schiesbn please have a look

**Comment by PVince81** [comment]
Critical because it makes downloads fail on some clients.
With the desktop sync client the first time the download fails, but the second time somehow it manages to finish it (no timeout), but the DB entry stays wrong.

I also noticed that the size propagation doesn't happen. Remember that whenever a file is uploaded, the parent folders is usually updated through `correctFolderSize`. Observing the database shows that it isn't happening. It only happens during the next PROPFIND or after a delay. Could be a separate issue.

**Comment by DeepDiver1975** [comment]
thx @PVince81

**Comment by PVince81** [comment]
Raised the folder size propagation issue here: https://github.com/owncloud/core/issues/18006
It's completely unrelated to this ticket, just happened to find it while investigating.

**Comment by schiesbn** [comment]
Hm, I could reproduce https://github.com/owncloud/core/issues/17594#issuecomment-126608468 once. Then I enabled the debugger step through it and it worked correctly. Since then it always worked also if I disabled the debugger again.... Maybe a race condition...

**Comment by PVince81** [comment]
Argh... in my local test cases it happened consistently. But as we know different envs react differently to race conditions.

But how could a single PUT request cause a race condition ? I didn't have any sync client running not web UI open.

**Comment by schiesbn** [comment]
Depends on how the scanner works. As described here https://github.com/owncloud/core/issues/17594#issuecomment-123297504 we remember the unencryted size directly after we finish writing the file (fclose). Then the scanner should start to index the file, the wrapper will detect the unenecrypted size and returns this together with "encrypted=true". If for some reasons the scanner scans the file before we set  the unencrypted size it could happen that both unencrypted size and the encrypted flag are wrong.

**Comment by PVince81** [comment]
@schiesbn tried again, happens consistently on my env.

I think there are two issues:
1) The issue that @xn--nding-jua had with concurrent scanner, this one can be "solved" by setting "filesystem_check_changes" to 0 and possibly with transactional locking later.

2) The issue I raised above that has nothing to do with concurrent. When following the steps from https://github.com/owncloud/core/issues/17594#issuecomment-126608468 and even with "filesystem_check_changes" set to 0, I still observe the issue where the "encrypted" flag is 0. Only focus on that for now, not the file size.

**Comment by schiesbn** [comment]
That's exactly what I tried

**Comment by PVince81** [comment]
Discussed on IRC with @schiesbn, the size and encrypted flag should be returned by `getMetadata`.

I debugged into scanFile from here https://github.com/owncloud/core/blob/master/lib/private/connector/sabre/file.php#L211

It then calls `getData()` here:
```
0  OC\Files\Cache\SharedScanner->getData() /srv/www/htdocs/owncloud/apps/files_sharing/lib/scanner.php:40
1  OC\Files\Cache\Scanner->scanFile() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:141
2  OC\Connector\Sabre\File->put() /srv/www/htdocs/owncloud/lib/private/connector/sabre/file.php:211
3  OC\Connector\Sabre\Directory->createFile() /srv/www/htdocs/owncloud/lib/private/connector/sabre/directory.php:113
4  Sabre\DAV\Server->createFile() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:1053
5  Sabre\DAV\CorePlugin->httpPut() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php:513
6  call_user_func_array:{/srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
7  Sabre\Event\EventEmitter->emit() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
8  Sabre\DAV\Server->invokeMethod() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:469
9  Sabre\DAV\Server->exec() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:254
10 require_once()  /srv/www/htdocs/owncloud/apps/files/appinfo/remote.php:83
11 {main}          /srv/www/htdocs/owncloud/remote.php:132
```

But for some strange reason it never reaches the encryption wrapper. So the return data doesn't contain any encryption-related information:
```
$data                            = (array[6]);
  $data['mimetype']              = (string[10]) 'text/plain';
  $data['mtime']                 = (int) '1438336587';
  $data['size']                  = (int) '14';
  $data['etag']                  = (string[32]) '5d2429b10c30bac434651a819afc87fe';
  $data['storage_mtime']         = (int) '1438336587';
  $data['permissions']           = (int) '27';
```

**Comment by PVince81** [comment]
This is the storage chain on which `getMetaData()` is called:
```
$this                            = (object|OC\Files\Cache\SharedScanner[7]);
  $this->storage                 = (object|OCA\Files_Trashbin\Storage[5]);
    $this->storage::disableTrash = (bool) '0';
    $this->storage->mountPoint   = (string[18]) '/user2/files/test/';
    $this->storage->deletedFiles = (array[0]);
    $this->storage->userManager  = (object|OC\User\Manager[4])+;
    $this->storage->storage      = (object|OC\Files\Storage\Shared[10]);
      $this->storage->storage::isInitialized = (array[0]);
      $this->storage->storage->share = (array[21])+;
      $this->storage->storage->files = (array[16])+;
      $this->storage->storage->ownerView = (object|OC\Files\View[4])+;
      $this->storage->storage->cache = (null);
      $this->storage->storage->scanner = (null);
      $this->storage->storage->watcher = (null);
      $this->storage->storage->storageCache = (null);
      $this->storage->storage->mountOptions = (array[0]);
      $this->storage->storage->cachedFiles = (array[0]);
  $this->storageId               = (string[13]) 'shared::/test';
  $this->cache                   = (object|OC\Files\Cache\Shared_Cache[7])+;
  $this->cacheActive             = (bool) '1';
  $this->useTransactions         = (bool) '1';
  $this->lockingProvider         = (object|OC\Lock\NoopLockingProvider[0]);
  $this->listeners               = (array[0]);
```

**Comment by PVince81** [comment]
@schiesbn can you compare with yours ?

**Comment by PVince81** [comment]
@schiesbn please debug to https://github.com/owncloud/core/issues/17594#issuecomment-126630797 then post the expanded storage chain from `$this->storage` and also `$sourceStorage`.

In my env `$this->storage` doesn't contain the encryption wrapper, but `$sourceStorage` does.
However the code only calls `$sourceStorage->getPermissions()` here https://github.com/owncloud/core/blob/stable8.1/apps/files_sharing/lib/scanner.php#L42

**Comment by PVince81** [comment]
Ok I put a breakpoint in the encryption wrapper's "getMetadata".
It's called several times for the part folder, and then when called on the final file, it's here:
```
0  OC\Files\Storage\Wrapper\Encryption->getMetaData() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:150
1  OC\Files\Storage\Wrapper\Wrapper->getMetaData() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:543
2  OC\Files\Cache\Scanner->getData() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:113
3  OC\Files\Cache\Scanner->scanFile() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:141
4  OC\Files\Cache\Scanner->scan() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:265
5  OC\Files\Cache\Updater->update() /srv/www/htdocs/owncloud/lib/private/files/cache/updater.php:114
6  OC\Files\View->rename() /srv/www/htdocs/owncloud/lib/private/files/view.php:697
7  OC\Encryption\Keys\Storage->renameKeys() /srv/www/htdocs/owncloud/lib/private/encryption/keys/storage.php:252
8  OC\Files\Storage\Wrapper\Encryption->rename() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:244
9  OCA\Files_Trashbin\Storage->rename() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/storage.php:77
10 OC\Files\Storage\Wrapper\Wrapper->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:532
11 OC\Files\Storage\Shared->rename() /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php:352
12 OCA\Files_Trashbin\Storage->rename() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/storage.php:77
13 OC\Files\Storage\Wrapper\Wrapper->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:532
14 OC\Connector\Sabre\File->put() /srv/www/htdocs/owncloud/lib/private/connector/sabre/file.php:191
15 OC\Connector\Sabre\Directory->createFile() /srv/www/htdocs/owncloud/lib/private/connector/sabre/directory.php:113
16 Sabre\DAV\Server->createFile() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:1053
17 Sabre\DAV\CorePlugin->httpPut() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php:513
18 call_user_func_array:{/srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
19 Sabre\Event\EventEmitter->emit() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
20 Sabre\DAV\Server->invokeMethod() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:469
21 Sabre\DAV\Server->exec() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:254
22 require_once()  /srv/www/htdocs/owncloud/apps/files/appinfo/remote.php:83
23 {main}          /srv/www/htdocs/owncloud/remote.php:132
```

The resulting `$data`:
```
$data                            = (array[6]);
  $data['mimetype']              = (string[20]) 'httpd/unix-directory';
  $data['mtime']                 = (int) '1438337587';
  $data['size']                  = (int) '-1';
  $data['etag']                  = (string[13]) '55bb4a3b7a856';
  $data['storage_mtime']         = (int) '1438337587';
  $data['permissions']           = (int) '31';
```

It doesn't enter any of the blocks that set the `encrypted` flag !

**Comment by PVince81** [comment]
More values from that specific call:
```
// by property_get
$path                            = (string[40]) 'files_encryption/keys/files/test/t14.txt';


// by property_get
$fullPath                        = (string[47]) '/user1/files_encryption/keys/files/test/t14.txt';

// => eval: $this->unencryptedSize
$evalResult                      = (array[2]);
  $evalResult->/user1/files/test/t14.txt.ocTransferId898383818.part = (int) '14';
  $evalResult->/user1/files/test/t14.txt = (int) '14';


// by property_get
$info                            = (bool) '0';
```

Ok, seems to be the key's path, not the actual file.

Continuing debugging...

**Comment by PVince81** [comment]
The encryption wrapper is called with the following paths, in sequence:
```
'/owncloud/remote.php/webdav/test/t15.txt';
'files_encryption/keys/files/test/t15.txt.ocTransferId28053470.part/OC_DEFAULT_MODULE';
'files_encryption/keys/files/test/t15.txt.ocTransferId28053470.part/OC_DEFAULT_MODULE/fileKey';
'files_encryption/keys/files/test/t15.txt.ocTransferId28053470.part/OC_DEFAULT_MODULE/user1.shareKey';
'files_encryption/keys/files/test/t15.txt.ocTransferId28053470.part/OC_DEFAULT_MODULE/user2.shareKey';
'files_encryption/keys/files/test/t15.txt';
'files_encryption/keys/files/test/t15.txt/OC_DEFAULT_MODULE';
```

It is never called for the actual final file.

**Comment by PVince81** [comment]
This is the place where it should write the size and flag: https://github.com/owncloud/core/blob/master/lib/private/files/storage/wrapper/encryption.php#L134

In my env it never reaches it.
It does call filesize here:
```
0  OC\Files\Storage\Wrapper\Encryption->filesize() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:128
1  OC\Files\Storage\Wrapper\Wrapper->filesize() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:140
2  OC\Files\Storage\Shared->filesize() /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php:219
3  OC\Files\Storage\Common->getMetaData() /srv/www/htdocs/owncloud/lib/private/files/storage/common.php:617
4  OC\Files\Storage\Wrapper\Wrapper->getMetaData() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:543
5  OC\Files\Cache\Scanner->getData() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:113
6  OC\Files\Cache\SharedScanner->getData() /srv/www/htdocs/owncloud/apps/files_sharing/lib/scanner.php:39
7  OC\Files\Cache\Scanner->scanFile() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:141
8  OC\Files\Cache\Scanner->scanChildren() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:344
9  OC\Files\Cache\Scanner->scan() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:267
10 OC\Files\Cache\Watcher->checkUpdate() /srv/www/htdocs/owncloud/lib/private/files/cache/watcher.php:91
11 OC\Files\Cache\Shared_Watcher->checkUpdate() /srv/www/htdocs/owncloud/apps/files_sharing/lib/watcher.php:41
12 OC\Files\View->getFileInfo() /srv/www/htdocs/owncloud/lib/private/files/view.php:1182
13 OC\Connector\Sabre\ObjectTree->getNodeForPath() /srv/www/htdocs/owncloud/lib/private/connector/sabre/objecttree.php:148
14 Sabre\DAV\Tree->nodeExists() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Tree.php:103
15 Sabre\DAV\CorePlugin->httpPut() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php:494
16 call_user_func_array:{/srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
17 Sabre\Event\EventEmitter->emit() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
18 Sabre\DAV\Server->invokeMethod() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:469
19 Sabre\DAV\Server->exec() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:254
20 require_once()  /srv/www/htdocs/owncloud/apps/files/appinfo/remote.php:83
21 {main}          /srv/www/htdocs/owncloud/remote.php:132
```

but the `$this->unencryptedSize` array is empty, so it never enters the block.

**Comment by schiesbn** [comment]
That's where the cache gets updated on my system:

![1](https://cloud.githubusercontent.com/assets/1589737/9005677/729841a6-3780-11e5-94f6-e670d0db5113.png)

**Comment by PVince81** [comment]
Here we go, almost the same stack:
```
0  OC\Files\Storage\Wrapper\Encryption->filesize() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:125
1  OC\Files\Storage\Wrapper\Wrapper->filesize() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:140
2  OC\Files\Storage\Shared->filesize() /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php:219
3  OC\Files\Storage\Common->getMetaData() /srv/www/htdocs/owncloud/lib/private/files/storage/common.php:617
4  OC\Files\Storage\Wrapper\Wrapper->getMetaData() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:543
5  OC\Files\Cache\Scanner->getData() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:113
6  OC\Files\Cache\SharedScanner->getData() /srv/www/htdocs/owncloud/apps/files_sharing/lib/scanner.php:39
7  OC\Files\Cache\Scanner->scanFile() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:141
8  OC\Connector\Sabre\File->put() /srv/www/htdocs/owncloud/lib/private/connector/sabre/file.php:211
9  OC\Connector\Sabre\Directory->createFile() /srv/www/htdocs/owncloud/lib/private/connector/sabre/directory.php:113
10 Sabre\DAV\Server->createFile() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:1053
11 Sabre\DAV\CorePlugin->httpPut() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php:513
12 call_user_func_array:{/srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
13 Sabre\Event\EventEmitter->emit() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
14 Sabre\DAV\Server->invokeMethod() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:469
15 Sabre\DAV\Server->exec() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:254
16 require_once()  /srv/www/htdocs/owncloud/apps/files/appinfo/remote.php:83
17 {main}          /srv/www/htdocs/owncloud/remote.php:132
```

Strangely your stack doesn't show `1  OC\Files\Storage\Wrapper\Wrapper->filesize() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:140`, but could be a IDE difference.

Here are the variables during this call:
```
$path                            = (string[17]) 'files/test/t2.txt';
$fullPath                        = (string[24]) '/user1/files/test/t2.txt';
$info                            = (bool) '0';
$this->unencryptedSize                      = (array[2]);
  $evalResult->/user1/files/test/t2.txt.ocTransferId1909706986.part = (int) '14';
  $evalResult->/user1/files/test/t2.txt = (int) '14';
$size                            = (int) '14';
```

It seems to expect that "fileid" is set. Since $info is empty, it doesn't store anything in the cache.
Also at this specific stage, there is no cache entry in the database for "t2.txt".

@schiesbn do you see such entry ?

Maybe there's a piece of code earlier that didn't properly create the cache entry that is expected to exist here.

**Comment by schiesbn** [comment]
Yes, for me the file is in the cache, that's why I have a fileid and the cache gets updated with the unenecrypted size and the encrypted flag. That's why I suggested earlier to try to remove the if condition and see if it works? Maybe the file cache can handle this case.

**Comment by PVince81** [comment]
I don't think it's correct to remove the condition. As an API consumer I don't expect a call to "filesize" to insert a cache entry when there is none. And there is a risk of causing further side effects.

We need to find out why the cache entry is missing.
What you could do is put a breakpoint in the cache code https://github.com/owncloud/core/blob/stable8.1/lib/private/files/cache/cache.php#L286 (maybe with a condition that `$path` contains the real file name) and find out the stack trace(s) when the cache entry is first inserted.

**Comment by schiesbn** [comment]
> I don't think it's correct to remove the condition. As an API consumer I don't expect a call to "filesize" to insert a cache entry when there is none. And there is a risk of causing further side effects.

If the file cache is just about to create the first version of the entry it also need to collect data like the filesize. So this would be a situation where you would end up in the method but without a file cache. If the file cache can handle partial data removing the if might be a solution (iirc I saw something like this in the past). Just give it a try if it works without a error message or not. Then we can still think about if this is the right way or not.

**Comment by PVince81** [comment]
Yes, it works when removing this condition.
This also means that the "put()" will first put a partial file cache entry containing only path, size and encrypted flag. The entry will even have no "parent" entry for a short time frame. This is why I'd like to have such incomplete data in the cache.

Let's see if we can find a proper solution.

**Comment by schiesbn** [comment]
> The entry will even have no "parent" entry for a short time frame. This is why I'd like to have such incomplete data in the cache.

Is this entry really in the database or is it just stored in a internal array to complete the data once the file cache gets written? In this case it wouldn't be a problem.

**Comment by PVince81** [comment]
Just checked it and you're right. When calling $this->cache->put() it doesn't store anything in the database but into a structure `$this->partial` (first time I saw this). And later on cache->get() will return that information. Strange behavior, but might make sense in some cases.

So it works... feel free to make a PR to remove that check in the meantime, that would be the quickfix for 8.1.1 in case we don't find the root cause. I still hope it won't have other side effects.

I'll try and find the root cause meanwhile.

**Comment by PVince81** [comment]
I have setup another computer (desktop) and am also seeing the issue.

Now I just found out that disabling the trashbin solved the issue.

Maybe the trashbin wrapper is messing up ? @schiesbn did you have trashbin disabled ? That could explain why your stack was missing a "Wrapper" entry.

**Comment by PVince81** [comment]
On all my envs, if I disable trashbin it works fine.

Next step: find out in what way the trashbin influences the outcome.

**Comment by PVince81** [comment]
Result so far:
- when trashbin is enabled, there is no cache entry before the call to scanFile()
- when trashbin is disabled, a cache entry exists with size=0 and encrypted=0

**Comment by PVince81** [comment]
Without encryption, the extra cache entry gets inserted here:
```
0  OC\Files\Cache\Cache->put() /srv/www/htdocs/owncloud/lib/private/files/cache/cache.php:307
1  OC\Files\Cache\Scanner->addToCache() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:225
2  OC\Files\Cache\Scanner->scanFile() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:188
3  OC\Files\Cache\Scanner->scan() /srv/www/htdocs/owncloud/lib/private/files/cache/scanner.php:265
4  OC\Files\View->getFileInfo() /srv/www/htdocs/owncloud/lib/private/files/view.php:1180
5  OC\Share\Share::getUsersSharingFile() /srv/www/htdocs/owncloud/lib/private/share/share.php:130
6  OCP\Share::getUsersSharingFile() /srv/www/htdocs/owncloud/lib/public/share.php:89
7  OC\Encryption\File->getAccessList() /srv/www/htdocs/owncloud/lib/private/encryption/file.php:57
8  OC\Files\Stream\Encryption->stream_open() /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:248
9  fopen()         /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:189
10 OC\Files\Stream\Encryption::wrapSource() /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:189
11 OC\Files\Stream\Encryption::wrap() /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:171
12 OC\Files\Storage\Wrapper\Encryption->fopen() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:426
13 OC\Files\Storage\Wrapper\Encryption->copyBetweenStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:516
14 OC\Files\Storage\Wrapper\Encryption->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:450
15 OC\Files\Storage\Shared->moveFromStorage() /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php:619
16 OC\Connector\Sabre\File->put() /srv/www/htdocs/owncloud/lib/private/connector/sabre/file.php:191
17 OC\Connector\Sabre\Directory->createFile() /srv/www/htdocs/owncloud/lib/private/connector/sabre/directory.php:113
18 Sabre\DAV\Server->createFile() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:1053
19 Sabre\DAV\CorePlugin->httpPut() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php:513
20 call_user_func_array:{/srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
21 Sabre\Event\EventEmitter->emit() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
22 Sabre\DAV\Server->invokeMethod() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:469
23 Sabre\DAV\Server->exec() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:254
24 require_once()  /srv/www/htdocs/owncloud/apps/files/appinfo/remote.php:83
25 {main}          /srv/www/htdocs/owncloud/remote.php:132
```

Now what does the trashbin do to block this ?

**Comment by PVince81** [comment]
Without trashbin, Encryption::stream_open() is called three times, with the following values for `$sharePath`:
- '/user1/files/test/t20.txt.ocTransferId380882296.part';
- '/user1/files/test/t20.txt.ocTransferId380882296.part';
- '/user1/files/test/t20.txt';

However with trashbin enabled, it's only called once on the part file.

Very strange.

**Comment by PVince81** [comment]
When trashbin is enabled, it does Encryption::rename instead of Encryption::copyFromStorage:
```
0  OC\Files\Storage\Local->rename() /srv/www/htdocs/owncloud/lib/private/files/storage/local.php:245
1  OC\Files\Storage\Wrapper\Encryption->rename() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:235
2  OCA\Files_Trashbin\Storage->rename() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/storage.php:77
3  OC\Files\Storage\Wrapper\Wrapper->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:532
4  OC\Files\Storage\Shared->rename() /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php:352
5  OCA\Files_Trashbin\Storage->rename() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/storage.php:77
6  OC\Files\Storage\Wrapper\Wrapper->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:532
7  OC\Connector\Sabre\File->put() /srv/www/htdocs/owncloud/lib/private/connector/sabre/file.php:191
8  OC\Connector\Sabre\Directory->createFile() /srv/www/htdocs/owncloud/lib/private/connector/sabre/directory.php:113
9  Sabre\DAV\Server->createFile() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:1053
10 Sabre\DAV\CorePlugin->httpPut() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php:513
11 call_user_func_array:{/srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
12 Sabre\Event\EventEmitter->emit() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
13 Sabre\DAV\Server->invokeMethod() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:469
14 Sabre\DAV\Server->exec() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:254
15 require_once()  /srv/www/htdocs/owncloud/apps/files/appinfo/remote.php:83
16 {main}          /srv/www/htdocs/owncloud/remote.php:132
```

I suspect that copyFromStorage (the working case) would automatically create the target cache entry while rename did not for some reason.

This means there might be a potential bug hiding in other use cases, the ones that trigger the "rename" route instead of "copyFromStorage".

**Comment by PVince81** [comment]
Okay, the reason for the two different behaviors is as follows:
Whenever trashbin is enabled, the `moveFromStorage` method will redirect the call to `rename()` in case it detects the same storage, see https://github.com/owncloud/core/blob/stable8.1/lib/private/files/storage/wrapper/wrapper.php#L531

But when trashbin is disabled, the `moveFromStorage` call directly reaches the shared storage's `moveFromStorage` call, which itself later reaches `Encryption::copyFromStorage` which seems to pre-create the cache entry.

- [ ] fix default storage wrapper to only forward the call of moveFromStorage, no logic
- [ ] investigate why the rename() code path doesn't product any cache entry
- [ ] optionally: remove fileid check as advised by @schiesbn here https://github.com/owncloud/core/issues/17594#issuecomment-126648426

**Comment by schiesbn** [comment]
I think https://github.com/owncloud/core/blob/stable8.1/lib/private/files/storage/wrapper/wrapper.php#L531 should just call ````$this->storage->moveFromStorage();```` like (almost) any other method in this class works to make sure that we go through all storage wrappers with the same method.

At the end you will reach this method https://github.com/owncloud/core/blob/stable8.1/lib/private/files/storage/common.php#L586 which does the same check. But here it is OK because now you are at the bottom and executed all storage wrappers.

**Comment by schiesbn** [comment]
> investigate why the rename() code path doesn't product any cache entry

I think it does produce a cache entry but in a different way. If the part file would have a complete entry with the encrypted flag it would probably work but we don't add part files to the cache that's why there is no existing entry during rename so he creates a new one, this lead to the situation described here https://github.com/owncloud/core/issues/17594#issuecomment-126651069

**Comment by PVince81** [comment]
Yes, so now we have two potential fixes for this. That would do for 8.1.1.

Still need to investigate the rest because I see future potential issues.

**Comment by PVince81** [comment]
@schiesbn do you see any other use case / code path where rename could be used and would potentially miss the encrypted flag due to "rename()" ?

**Comment by schiesbn** [comment]
If it is a regular rename the file cache will copy the encryption flag from the source. The problem here is that the source is a part file -> no file cache entry -> nothing to copy for the file cache -> the cache collects the data from scratch -> lead to the situation we have.

I think we shoud do both, remove the if statement in the storage wrapper because the file cache handles this partial data correctly and change the wrapper to just forward the moveFromStorage() call. I will prepare two separate pull requests for both. Then we can discuss there and also see if all tests pass, etc.

**Comment by PVince81** [comment]
@schiesbn cool, thanks. I tried the moveFromStorage fix locally and it worked, so I agree to have this in.

**Comment by PVince81** [comment]
The fix we chose for 8.1.1 will be https://github.com/owncloud/core/pull/18009#issuecomment-126697678 as it's the less risky of both.

**Comment by PVince81** [comment]
Would be good to have a repair script to repair those broken files, raised here: https://github.com/owncloud/core/issues/18225

**Comment by xn--nding-jua** [comment]
Hello,

here is my proposal for a recovery-script. After the configuration is done in the header of the file it will do the following:

1. connect to the ownCloud encryption system
2. connect to the mySQL database
3. steps through the folder given in the header (recursively with sub-folders)
4. compares the filesize stored in database with the real encrypted file. If sizes are the same, the database has the wrong value
5. if the filesize is wrong (database===encrypted file), the specific file will be decrypted, stored to the tmp-directory and the filesize of the decrypted file will be stored to the database. The tmp-file will be deleted afterwards.

Be careful:
a) script will not test, if the encryption-app is enabled or not
b) will run only with Encryption App 2.0
c) database will be read for each file independently. maybe someone can optimize it


How to use?
1. copy the code into a file called fixwrongsizedfiles.php
2. edit the header of the file (usernames, passwords, folder, etc)
3. connect to server by ssh and run "php fixwrongsizedfiles.php"
4. the script will run over all files within the scope of the given folder and updates the SQL-DB whenever necessary


```
<?php

// set php-language to German/UTF-8
$loc_de = setlocale(LC_ALL, "de_DE.UTF-8");

// ownCloud setup
$oc_datadir = '/owncloud/data/';
$oc_tmpdir = '/tmp/';
$oc_username = "User1";
$oc_password = "Trustno1";
$db_server = "localhost";
$db_name = "DBNAME";
$db_password = "Trustno1";

// check all files within the following directory (with sub-directories)
// enter the folder as you see them within owncloud-webui
$searchfolder='Folder1/Subfolder1';


require_once 'lib/base.php';

////////////////////////// ////////////////////////////////
// Open the desired file and decrypt it to the tmp-directory
///////////////////////////////////////////////////////////

// first get users private key and decrypt it
try{
	$session = new \OCA\Encryption\Session(\OC::$server->getSession());
	$userSession = \OC::$server->getUserSession();
	$crypt = new \OCA\Encryption\Crypto\Crypt(\OC::$server->getLogger(), $userSession, \OC::$server->getConfig());

	$encryptedUserKey = file_get_contents($oc_datadir . $oc_username . '/files_encryption/OC_DEFAULT_MODULE/' . $oc_username . '.privateKey');
	$decryptedUserKey = $crypt->decryptPrivateKey($encryptedUserKey, $oc_password);
} catch (\Exception $ex) {
	echo $ex->getMessage() . PHP_EOL;
}

try{
	// connect to mySQL
	$db = mysql_connect($db_server, $db_name, $db_password);
	if(!$db)
	{
		// no connection to db
		echo "ERROR: Failed to connect to DB: " . mysqli_connect_error() . PHP_EOL;
	}else{
		// select database
		mysql_select_db($db_name) or die("ERROR: Failed to switch to db " . $db_name);

		//$folderhandle=opendir(utf8_encode($oc_datadir . $oc_username . "/files/" . $searchfolder));
		//while (false !== ($file = readdir($folderhandle)))
		$di = new RecursiveDirectoryIterator(utf8_encode($oc_datadir . $oc_username . "/files/" . $searchfolder));
		foreach (new RecursiveIteratorIterator($di) as $file)
		{
			// define the file to check
			if(file_exists($file) && (filetype($file)==="file"))
			{
				$oc_filepath = substr($file, strpos($file, utf8_encode($searchfolder)));
				$oc_filepath_raw = utf8_decode($oc_filepath);

				echo $oc_filepath . PHP_EOL;
				
				$dbresult = mysql_query("SELECT * FROM `" . $db_name .  "`.`oc_filecache` WHERE `path` LIKE 'files/" . $oc_filepath_raw . "'");
				$numberofrows = mysql_num_rows($dbresult);

				if($numberofrows>1)
				{
					// more than one entry found -> exception
					echo "ERROR: More than one DB-entry found for the file files/" . $oc_filepath . PHP_EOL;
				}elseif($numberofrows===0){
					// more than one entry found -> exception
					echo "ERROR: No DB-entry found for the file files/" . $oc_filepath . PHP_EOL;
				}else{
					// only one entry found -> OK
					$row = mysql_fetch_object($dbresult);


					$encryptedFilesize = filesize($oc_datadir . $oc_username . '/files/' . $oc_filepath);

					// compare encrypted file and stored filesize. if they match, we have to fix it
					if(intval($row->size)===intval($encryptedFilesize)) {
						echo "INFO: Filesize matches encrypted filesize. So we have to fix it." . PHP_EOL;


						// we need to decrypt the file-key, therefore we use our private key and the share key
						$shareKey = file_get_contents($oc_datadir . $oc_username . '/files_encryption/keys/files/' . $oc_filepath . '/OC_DEFAULT_MODULE/' . $oc_username . '.shareKey');
						$encryptedKeyfile = file_get_contents($oc_datadir . $oc_username . '/files_encryption/keys/files/' . $oc_filepath . '/OC_DEFAULT_MODULE/fileKey');
						try {
							$decryptedKeyfile = $crypt->multiKeyDecrypt($encryptedKeyfile, $shareKey, $decryptedUserKey);
						} catch (\Exception $ex) {
							echo $ex->getMessage() . PHP_EOL;
						}

						// read and decrypt the encrypted file
						$filehandle = fopen($oc_datadir . $oc_username . '/files/' . $oc_filepath, 'r');
						$encryptedContent = '';
						$decryptedContent = '';

						// read header (means removing it for decrypting the file)
						$encryptedContent = fread($filehandle, 8192);
						// read data
						while ($encryptedContent = fread($filehandle, 8192)) {
							// Decrypt the block
							$decryptedBlock = $crypt->symmetricDecryptFileContent($encryptedContent, $decryptedKeyfile);
							// A that point you can write the decrypted block to the decrypted file
							// instead of appending it to $decryptedContent. But I don't fear memory
							// limits and I choose to append the block to my string and write the
							//  file at the end
							$decryptedContent .= $decryptedBlock;
						}
						fclose($filehandle);

						// write the content to the tmp-folder
						// delete an existing temporary file
						if (file_exists($oc_tmpdir . basename($oc_filepath)))
						{
							unlink($oc_tmpdir . basename($oc_filepath));
						}
						file_put_contents($oc_tmpdir . basename($oc_filepath), $decryptedContent);

						// read the real filesize of the file
						$decryptedFilesize = filesize($oc_tmpdir . basename($oc_filepath));
						//echo "Size of encrypted file: " . $encryptedFilesize . " Byte" . PHP_EOL;
						//echo "Size of decrypted file: " . $decryptedFilesize . " Byte" . PHP_EOL;

						// delete the temporary file
						if (file_exists($oc_tmpdir . basename($oc_filepath)))
						{
							unlink($oc_tmpdir . basename($oc_filepath));
						}

						// update DB-entry
						if(intval($row->size)===intval($decryptedFilesize))
						{
							// stored size is the expected size of the dencrypted file -> nothing to do
							echo "INFO: Correct filesize of " . $row->size . " Bytes found in DB. Nothing to do here." . PHP_EOL;
						}elseif(intval($row->size)===intval($encryptedFilesize))
						{
							// stored size is the expected size of the encrypted file -> correct the DB-entry
							echo "INFO: Wrong filesize of " . $row->size . " Bytes found in DB. Will be fixed to " . $decryptedFilesize . " Bytes" . PHP_EOL;

							$update = mysql_query("UPDATE `" . $db_name .  "`.`oc_filecache` Set size = '" . $decryptedFilesize . "' WHERE `path` LIKE 'files/" . $oc_filepath_raw . "'");
						}else{
							// size is not as expected -> exception
							echo "ERROR: The filesize stored in the DB (" . $row->size . " Bytes) is not as expected (" . $encryptedFilesize . " Bytes)!" . PHP_EOL;
						}
					}else{
						echo "INFO: Filesize seems to be OK. Nothing to do here." . PHP_EOL;
					}
				}
			}
		}
		closedir($folderhandle);
		mysql_close($db);
	}
} catch (\Exception $ex) {
	echo $ex->getMessage() . PHP_EOL;
}

?>
```
